### PR TITLE
[FLINK-17209][sql-client] Allow users to specify dialect in sql-clien…

### PR DESCRIPTION
**Pull request by lirui-apache**
…t yaml

## What is the purpose of the change

Allow users to specify dialect in sql-client yaml.

## Brief change log

  - Add dialect property to `ExecutionEntry`, and create TableConfig using the specified dialect.
  - Add test case

## Verifying this change

Added test case.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? yes
  - If yes, how is the feature documented? docs

### CODE REVIEW

**Comment by lirui-apache** [comment]
I changed the implementations of dialect APIs to use a `ConfigOption`. I didn't change the API itself because changing the public API needs more formal discussion. This way user can set the dialect with either the API or configuration.
@JingsongLi Do you think it makes sense?

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/config/TableConfigOptions.java [hunk 1]
```diff
@@ -40,4 +41,12 @@ private TableConfigOptions() {}
 					.withDescription("Enable or disable the OPTIONS hint used to specify table options" +
 							"dynamically, if disabled, an exception would be thrown " +
 							"if any OPTIONS hint is specified");
+
+	@Documentation.TableOption(execMode = Documentation.ExecMode.BATCH_STREAMING)
+	public static final ConfigOption<String> TABLE_SQL_DIALECT = key("table.sql.dialect")
```
Comment (JingsongLi) [inline flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/config/TableConfigOptions.java:46]: table.sql-dialect?

File: flink-table/flink-sql-client/src/main/java/org/apache/flink/table/client/config/entries/ExecutionEntry.java [hunk 2]
```diff
@@ -110,6 +111,8 @@
 
 	public static final String EXECUTION_CURRENT_DATABASE = "current-database";
 
+	public static final String EXECUTION_SQL_DIALECT = "sql-dialect";
```
Comment (JingsongLi) [inline flink-table/flink-sql-client/src/main/java/org/apache/flink/table/client/config/entries/ExecutionEntry.java:114]: just `dialect`? it is sql-client

**Comment by lirui-apache** [comment]
@JingsongLi Thanks for the review. PR updated accordingly.

**Pull request merged**